packager/packager.cc

```diff
@@ -231,6 +267,112 @@ ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameR
+// prefix.
+class EnforcePackagePrefix {
+    const PackageInfo *pkg;
+    vector<core::NameRef> nameParts;
+    int rootConsts = 0;
+    int skipPush = 0;
+
+public:
+    EnforcePackagePrefix(const PackageInfo *pkg) : pkg(pkg) {
+        ENFORCE(pkg != nullptr);
+    }
+
+    ast::ExpressionPtr preTransformClassDef(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &classDef = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+        if (classDef.symbol == core::Symbols::root()) {
+            // Ignore top-level <root>
+            return tree;
+        }
+        const auto &pkgName = pkg->name.fullName.parts;
+        if (nameParts.size() > pkgName.size()) {
+            // At this depth we can stop checking the prefixes since beyond the end of the prefix.
+            skipPush++;
+        } else {
+            auto &constantLit = ast::cast_tree_nonnull<ast::UnresolvedConstantLit>(classDef.name);
+            pushConstantLit(&constantLit);
+
+            size_t minSize = std::min(pkgName.size(), nameParts.size());
+            if (rootConsts == 0 && !std::equal(pkgName.begin(), pkgName.begin() + minSize, nameParts.begin(),
+                                               nameParts.begin() + minSize)) {
+                if (auto e = ctx.beginError(constantLit.loc, core::errors::Packager::DefinitionPackageMismatch)) {
+                    e.setHeader("Class or method definition must match enclosing package namespace `{}`",
+                                fmt::map_join(pkgName.begin(), pkgName.end(),
+                                              "::", [&](const auto &nr) { return nr.show(ctx); }));
+                }
+            }
+        }
+        return tree;
+    }
+
+    ast::ExpressionPtr postTransformClassDef(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &classDef = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+        if (classDef.symbol == core::Symbols::root()) {
+            // Sanity check bookkeeping
+            ENFORCE(nameParts.size() == 0);
+            ENFORCE(rootConsts == 0);
+            ENFORCE(skipPush == 0);
+            return tree;
+        }
+        auto *constantLit = &ast::cast_tree_nonnull<ast::UnresolvedConstantLit>(classDef.name);
+        if (skipPush > 0) {
+            skipPush--;
+        } else {
+            popConstantLit(constantLit);
+        }
+        return tree;
+    }
+
+    ast::ExpressionPtr preTransformAssign(core::MutableContext ctx, ast::ExpressionPtr original) {
+        auto &asgn = ast::cast_tree_nonnull<ast::Assign>(original);
+        auto lhs = ast::cast_tree<ast::UnresolvedConstantLit>(asgn.lhs);
+        if (lhs != nullptr) {
+            auto &pkgName = pkg->name.fullName.parts;
+            size_t minSize = std::min(pkgName.size(), nameParts.size());
+            if (rootConsts == 0 &&
+                !std::equal(pkgName.begin(), pkgName.end(), nameParts.begin(), nameParts.begin() + minSize)) {
```

Comment:
You don't want minSize in that spot because that would allow for the definition:

```
class Foo
  X = 'something'
end
```
Inside code that was meant to be packaged inside `Foo::Bar`. (min size = 1, so we'd only compare that `[Foo] == [Foo]`

Reply:
I'll factor into funcs with more descriptive name to highlight the subtle difference.